os/dnd.h

```diff
@@ -0,0 +1,135 @@
+os::SurfaceRef default_decode_bmp(const uint8_t* buf, const uint32_t len);
+os::SurfaceRef default_decode_gif(const uint8_t* buf, const uint32_t len);
+
+namespace os {
+
+  class Window;
+
+#if CLIP_ENABLE_IMAGE
+  using DecoderFunc = std::function<SurfaceRef(const uint8_t* buf, const uint32_t len)>;
+
+  // Methods used to configure custom decoder functions by replacing the default implementations.
+
+  void set_decode_png(DecoderFunc func);
+  void set_decode_jpg(DecoderFunc func);
+  void set_decode_bmp(DecoderFunc func);
+  void set_decode_gif(DecoderFunc func);
+
+  SurfaceRef decode_png(const uint8_t* buf, const uint32_t len);
+  SurfaceRef decode_jpg(const uint8_t* buf, const uint32_t len);
+  SurfaceRef decode_bmp(const uint8_t* buf, const uint32_t len);
+  SurfaceRef decode_gif(const uint8_t* buf, const uint32_t len);
+#endif
+  // Operations that can be supported by source and target windows in a drag
+  // and drop operation.
+  enum class DropOperation {
+    None = 0,
+    Copy = 1,
+    Move = 2,
+    Link = 4,
+    Any = Copy | Move | Link,
+  };
+
+  // Types of representations supported for each DragDataItem.
+  enum class DragDataItemType {
+    Paths,
+    Image,
+    Url,
+  };
+
+  // Interface to get dragged data from the platform's implementation.
+  class DragDataProvider {
+  public:
+    virtual ~DragDataProvider() {}
+    virtual base::paths getPaths() = 0;
+#if CLIP_ENABLE_IMAGE
+    virtual SurfaceRef getImage() = 0;
+#endif
+    virtual std::string getUrl() = 0;
+    virtual bool contains(DragDataItemType type) { return false; }
+  };
+
+  class DragEvent {
+  public:
+    DragEvent(os::Window* target,
+              DropOperation supportedOperations,
+              const gfx::Point& dragPosition,
+              DragDataProvider* dataProvider)
+      : m_target(target)
+      , m_supportedOperations(supportedOperations)
+      , m_position(dragPosition)
+      , m_dataProvider(dataProvider) {}
+
+    // Destination window of the DragEvent.
+    os::Window* target() const { return m_target; }
+    DropOperation dropResult() const { return m_dropResult; }
+    DropOperation supportedOperations() const { return m_supportedOperations; }
+    bool acceptDrop() const { return m_acceptDrop; }
+    const gfx::Point& position() { return m_position; }
+    DragDataProvider* dataProvider() { return m_dataProvider; }
+
+    // Sets what will be the outcome of dropping the dragged data when it is
+    // accepted by the target window. Only one of the enum values should be passed,
+    // do not combine values using bitwise operators.
+    void dropResult(DropOperation operation) { m_dropResult = operation; }
+    // Set this to true when the dropped data was accepted/processed by the
+    // target window, or set to false otherwise.
+    void acceptDrop(bool value) { m_acceptDrop = value; }
+
+    bool sourceSupports(DropOperation op) {
+      return static_cast<int>(m_supportedOperations) & static_cast<int>(op);
```

Comment:
Why is it needed? I want to return true if `static_cast<int>(m_supportedOperations) & static_cast<int>(op)` is not zero, which should be only true if the bits from `m_supportedOperations` ANDed with the bits of `op` gives a value greater than zero. Or is this just a clang-tidy thing?

Reply:
It depends on how `sourceSupports()` should work if we specify several flags. For example I was thinking `sourceSupports(Copy | Link)` should return true if the source supports both operations (Copy+Link, so this change is required). But now I can imagine something like `sourceSupports(Any)` and I'm confused, because it should return true if any operation is supported (as it's now in the PR), but I was thinking about `All` instead of `Any` (returning true if all operation are supported).

So I'm not sure if `DropOperation::Any` should be called `DropOperation::All` or probably we should remove this constant to avoid confusion, or keep it as `Any` rename it as `All` depending on what is expected in all functions that receive/return these flags.

Reply:
I see. The semantics of `sourceSupports(Copy | Link)` is that it returns true if it supports Copy OR Link (or both maybe). Since you are using a bitwise OR, it actually makes sense to me to use it in this way. And the same reasoning was applied to name the DropOperation `Any`, because the function returns true if it supports any of the indicated flags (not all).

If you want to change the behavior of this function to work as you suggested, then yes, the name `All` would be appropriate in this case. Let me know what you want to do.

Reply:
Probably I'd prefer to change the behavior with the `== static_cast<int>(op)` and rename to `All` option. We're already doing exactly this in Aseprite in `doc::Layer::hasFlags()` and `ui::Widget::hasFlags()`.

This is because a bitwise OR works like an `a || b` only if the operands are the same bit, but here we generally combine different bits/flags, in this case a bitwise OR is like an addition of the values (a+b).

Reply:
I know how bitwise OR works, no need to explain it, thanks anyway. As I said, the current behavior was intentionally done that way. However I'm happy to change it because it is what you prefer. Thank you.

Reply:
Just in case I'm not lecturing you how bitwise OR works, I'm explaining the logic, or the though process we could say, behind these "hasFlags" functions I had at the moment and why I think it makes sense to works in this way.

> because it is what you prefer

In case you think this is wrong or is a better way, you can just propose that better way. For example, one alternative would be to have functions like `supportsAny()` or `supportsAll()`, which makes no mistake on the intention from the client side.

Reply:
I have already done what I think it is the best way. It didn't match your expectations, that's fine, there is not much else to do on my side besides doing what you expected. I'm not saying that what you say is wrong, it is just a different approach which also works.

Reply:
Also, I prefer to have just one function for this, it doesn't worth it to have two for almost the same.

Reply:
> EDIT: Sorry, you was saying one or the other...it could be a possibility to name it supportsAny, it makes sense...but now that I know what you expected, I would hardly try to make it as I thought. Thank you anyway.

I just saw your edit. No problem about this, I prefer the shorter version too (`supports()`) and in most cases only one operation will be used (Copy)